controllers/dockermachine_controller.go

```diff
@@ -91,137 +132,138 @@ func (r *DockerMachineReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
 	// Always attempt to Patch the DockerMachine object and status after each reconciliation.
 	defer func() {
 		if err := patchHelper.Patch(ctx, dockerMachine); err != nil {
-			log.Error(err, "failed to patch DockerCluster")
-			if reterr == nil {
-				reterr = err
+			log.Error(err, "failed to patch DockerMachine")
+			if rerr == nil {
+				rerr = err
 			}
 		}
 	}()
 
+	// Handle deleted machines
+	if !dockerMachine.ObjectMeta.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(machine, dockerMachine, externalMachine, externalLoadBalancer)
+	}
+
+	// Handle non-deleted machines
+	return r.reconcileNormal(machine, dockerMachine, externalMachine, externalLoadBalancer)
+}
+
+func (r *DockerMachineReconciler) reconcileNormal(machine *clusterv1.Machine, dockerMachine *infrav1.DockerMachine, externalMachine *docker.Machine, externalLoadBalancer *docker.LoadBalancer) (ctrl.Result, error) {
 	// If the DockerMachine doesn't have finalizer, add it.
 	if !util.Contains(dockerMachine.Finalizers, infrav1.MachineFinalizer) {
 		dockerMachine.Finalizers = append(dockerMachine.Finalizers, infrav1.MachineFinalizer)
 	}
-	state := getState(machine, dockerMachine)
-	log = log.WithValues("state", state.String())
-	// TODO: consider setting the key pieces in this function, such as ProviderID and deleting finalizers
-	switch state {
-	case Deleted:
-		log.Info("Start reconcileDelete dockerMachine")
-		return r.reconcileDelete(ctx, cluster, machine, dockerMachine)
-	case Provisioned:
-		return ctrl.Result{}, nil
-	case Pending:
-		log.Info("Waiting for machine bootstrap")
+
+	// if the machine is already provisioned, return
+	if dockerMachine.Spec.ProviderID != nil {
 		return ctrl.Result{}, nil
-	case Provisioning:
-		log.Info("Provisioning machine")
-		// Ensuring cluster is ready for joining
-		clusterExists, err := kindcluster.IsKnown(cluster.Name)
-		if err != nil {
-			log.Error(err, "Error finding cluster-name", "cluster", cluster.Name)
-			return ctrl.Result{}, err
-		}
+	}
 
-		// If there's no cluster, requeue the request until there is one
-		if !clusterExists {
-			r.Log.Info("There is no cluster yet, waiting for a cluster before creating machines")
-			return ctrl.Result{RequeueAfter: time.Second * 30}, nil
-		}
-		log.Info("Creating machine")
-		return r.create(ctx, cluster, machine, dockerMachine)
-	default:
-		log.Info("Unknown state", "state", state)
+	// Make sure bootstrap data is available and populated.
+	if machine.Spec.Bootstrap.Data == nil {
+		externalMachine.Info("Waiting for the Bootstrap provider controller to set bootstrap data")
```

Comment:
this should use the reconciler's logger, not the externalMachine logger since we're still in the reconcile function. It would be confusing to have a log line saying it came from the external machine when it actually came from reconcile